File: android/app/src/main/java/com/musicflow/app/PlaybackService.java
package com.musicflow.app;

import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.app.Service;
import android.content.Intent;
import android.media.AudioAttributes;
import android.media.AudioFocusRequest;
import android.media.AudioManager;
import android.media.MediaPlayer;
import android.media.session.MediaSession;
import android.media.session.PlaybackState;
import android.os.Build;
import android.os.Handler;
import android.os.IBinder;
import android.os.Looper;

/**
 * 媒体播放前台服务 — targetSdk 36 后台保活核心
 * 使用平台 API（无第三方依赖），解决锁屏断播、后台被杀
 */
public class PlaybackService extends Service
        implements MediaPlayer.OnCompletionListener, MediaPlayer.OnErrorListener,
        MediaPlayer.OnPreparedListener, AudioManager.OnAudioFocusChangeListener {

    private MediaPlayer mediaPlayer;
    private MediaSession mediaSession;
    private AudioManager audioManager;
    private AudioFocusRequest focusRequest;
    private Handler handler = new Handler(Looper.getMainLooper());
    private boolean isForeground = false;

    private String currentTitle = "MusicFlow";
    private String currentArtist = "";

    public static PlaybackService instance;

    private static final String CHANNEL_ID = "musicflow_playback";
    private static final int NOTIFICATION_ID = 1001;

    @Override
    public void onCreate() {
        super.onCreate();
        instance = this;
        audioManager = (AudioManager) getSystemService(AUDIO_SERVICE);
        createNotificationChannel();
        createMediaSession();
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        if (intent != null) {
            String action = intent.getStringExtra("action");
            if ("play".equals(action)) {
                playMedia(intent.getStringExtra("url"),
                        intent.getStringExtra("title"),
                        intent.getStringExtra("artist"));
            } else if ("pause".equals(action)) {
                pause();
            } else if ("resume".equals(action)) {
                resume();
            } else if ("stop".equals(action)) {
                stop();
            }
        }
        return START_STICKY;
    }

    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }

    // ============== 播放控制 ==============

    public void playMedia(String url, String title, String artist) {
        if (url == null || url.isEmpty()) return;
        currentTitle = title != null ? title : "未知歌曲";
        currentArtist = artist != null ? artist : "未知艺术家";
        releasePlayer();

        mediaPlayer = new MediaPlayer();
        mediaPlayer.setAudioAttributes(new AudioAttributes.Builder()
                .setContentType(AudioAttributes.CONTENT_TYPE_MUSIC)
                .setUsage(AudioAttributes.USAGE_MEDIA).build());
        mediaPlayer.setOnCompletionListener(this);
        mediaPlayer.setOnErrorListener(this);
        mediaPlayer.setOnPreparedListener(this);

        try {
            mediaPlayer.setDataSource(url);
            mediaPlayer.prepareAsync();
            updatePlaybackState(PlaybackState.STATE_BUFFERING);
        } catch (Exception e) {
            e.printStackTrace();
            stopSelf();
        }
    }

    @Override
    public void onPrepared(MediaPlayer mp) {
        requestAudioFocus();
        mp.start();
        updatePlaybackState(PlaybackState.STATE_PLAYING);
        startForegroundNotification();
    }

    public void pause() {
        if (mediaPlayer != null && mediaPlayer.isPlaying()) {
            mediaPlayer.pause();
            updatePlaybackState(PlaybackState.STATE_PAUSED);
            abandonAudioFocus();
            stopForegroundNotification();
        }
    }

    public void resume() {
        if (mediaPlayer != null && !mediaPlayer.isPlaying()) {
            requestAudioFocus();
            mediaPlayer.start();
            updatePlaybackState(PlaybackState.STATE_PLAYING);
            startForegroundNotification();
        }
    }

    public void stop() {
        releasePlayer();
        abandonAudioFocus();
        stopForegroundNotification();
        stopSelf();
    }

    public void seekTo(long ms) {
        if (mediaPlayer != null) mediaPlayer.seekTo((int) ms);
    }

    public long getCurrentPosition() {
        return mediaPlayer != null ? mediaPlayer.getCurrentPosition() : 0;
    }

    public long getDuration() {
        return mediaPlayer != null ? mediaPlayer.getDuration() : 0;
    }

    public boolean isPlaying() {
        return mediaPlayer != null && mediaPlayer.isPlaying();
    }

    // ============== 音频焦点 ==============

    private void requestAudioFocus() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            focusRequest = new AudioFocusRequest.Builder(AudioManager.AUDIOFOCUS_GAIN)
                    .setAudioAttributes(new AudioAttributes.Builder()
                            .setContentType(AudioAttributes.CONTENT_TYPE_MUSIC)
                            .setUsage(AudioAttributes.USAGE_MEDIA).build())
                    .setOnAudioFocusChangeListener(this, handler)
                    .build();
            audioManager.requestAudioFocus(focusRequest);
        } else {
            audioManager.requestAudioFocus(this, AudioManager.STREAM_MUSIC,
                    AudioManager.AUDIOFOCUS_GAIN);
        }
    }

    private void abandonAudioFocus() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O && focusRequest != null) {
            audioManager.abandonAudioFocusRequest(focusRequest);
        } else {
            audioManager.abandonAudioFocus(this);
        }
    }

    @Override
    public void onAudioFocusChange(int focusChange) {
        switch (focusChange) {
            case AudioManager.AUDIOFOCUS_LOSS:
                stop();
                break;
            case AudioManager.AUDIOFOCUS_LOSS_TRANSIENT:
                pause();
                break;
            case AudioManager.AUDIOFOCUS_LOSS_TRANSIENT_CAN_DUCK:
                if (mediaPlayer != null) mediaPlayer.setVolume(0.3f, 0.3f);
                break;
            case AudioManager.AUDIOFOCUS_GAIN:
                if (mediaPlayer != null) {
                    mediaPlayer.setVolume(1.0f, 1.0f);
                    if (!mediaPlayer.isPlaying()) {
                        mediaPlayer.start();
                        updatePlaybackState(PlaybackState.STATE_PLAYING);
                        startForegroundNotification();
                    }
                }
                break;
        }
    }

    // ============== MediaSession ==============

    private void createMediaSession() {
        mediaSession = new MediaSession(this, "MusicFlow");
        mediaSession.setCallback(new MediaSession.Callback() {
            @Override
            public void onPlay() { resume(); }
            @Override
            public void onPause() { pause(); }
            @Override
            public void onSeekTo(long pos) { seekTo(pos); }
            @Override
            public void onStop() { stop(); }
        });
        mediaSession.setActive(true);
    }

    private void updatePlaybackState(int state) {
        if (mediaSession == null) return;
        PlaybackState.Builder builder = new PlaybackState.Builder()
                .setState(state, getCurrentPosition(), 1.0f)
                .setActions(PlaybackState.ACTION_PLAY |
                        PlaybackState.ACTION_PAUSE |
                        PlaybackState.ACTION_SEEK_TO |
                        PlaybackState.ACTION_STOP |
                        PlaybackState.ACTION_SKIP_TO_NEXT |
                        PlaybackState.ACTION_SKIP_TO_PREVIOUS);
        mediaSession.setPlaybackState(builder.build());
    }

    // ============== 通知 ==============

    private void createNotificationChannel() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            NotificationChannel channel = new NotificationChannel(
                    CHANNEL_ID, "音乐播放", NotificationManager.IMPORTANCE_HIGH);
            channel.setLockscreenVisibility(Notification.VISIBILITY_PUBLIC);
            channel.setShowBadge(false);
            NotificationManager nm = (NotificationManager) getSystemService(NOTIFICATION_SERVICE);
            if (nm != null) nm.createNotificationChannel(channel);
        }
    }

    private void startForegroundNotification() {
        if (isForeground) return;
        isForeground = true;

        Intent intent = new Intent(this, MainActivity.class);
        intent.setFlags(Intent.FLAG_ACTIVITY_SINGLE_TOP);
        PendingIntent pi = PendingIntent.getActivity(this, 0, intent,
                PendingIntent.FLAG_UPDATE_CURRENT | PendingIntent.FLAG_IMMUTABLE);

        Notification.Builder nb;
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            nb = new Notification.Builder(this, CHANNEL_ID);
        } else {
            nb = new Notification.Builder(this);
        }

        Notification notif = nb.setContentTitle(currentTitle)
                .setContentText(currentArtist)
                .setSmallIcon(android.R.drawable.ic_media_play)
                .setContentIntent(pi)
                .setOngoing(true)
                .setPriority(Notification.PRIORITY_HIGH)
                .setStyle(new Notification.MediaStyle()
                        .setMediaSession(mediaSession.getSessionToken())
                        .setShowActionsInCompactView(0, 1, 2))
                .addAction(android.R.drawable.ic_media_previous, "上一首", null)
                .addAction(android.R.drawable.ic_media_pause, "暂停", null)
                .addAction(android.R.drawable.ic_media_next, "下一首", null)
                .build();

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            startForeground(NOTIFICATION_ID, notif,
                    android.content.pm.ServiceInfo.FOREGROUND_SERVICE_TYPE_MEDIA_PLAYBACK);
        } else {
            startForeground(NOTIFICATION_ID, notif);
        }
    }

    private void stopForegroundNotification() {
        if (!isForeground) return;
        isForeground = false;
        stopForeground(true);
    }

    // ============== 播放器回调 ==============

    @Override
    public void onCompletion(MediaPlayer mp) {
        updatePlaybackState(PlaybackState.STATE_STOPPED);
        abandonAudioFocus();
        stopForegroundNotification();
    }

    @Override
    public boolean onError(MediaPlayer mp, int what, int extra) {
        releasePlayer();
        abandonAudioFocus();
        stopForegroundNotification();
        stopSelf();
        return true;
    }

    // ============== 生命周期 ==============

    private void releasePlayer() {
        if (mediaPlayer != null) {
            try { if (mediaPlayer.isPlaying()) mediaPlayer.stop(); } catch (Exception ignored) {}
            mediaPlayer.release();
            mediaPlayer = null;
        }
    }

    @Override
    public void onTaskRemoved(Intent rootIntent) {
        if (mediaPlayer == null || !mediaPlayer.isPlaying()) {
            stopSelf();
        }
    }

    @Override
    public void onDestroy() {
        instance = null;
        releasePlayer();
        abandonAudioFocus();
        if (mediaSession != null) {
            mediaSession.setActive(false);
            mediaSession.release();
            mediaSession = null;
        }
        super.onDestroy();
    }
}

File: android/app/src/main/java/com/musicflow/app/PlaybackService.kt
package com.musicflow.app

import android.app.PendingIntent
import android.content.Intent
import android.content.pm.ServiceInfo
import android.os.Build
import android.os.Bundle
import android.support.v4.media.MediaBrowserCompat
import android.support.v4.media.session.MediaSessionCompat
import android.support.v4.media.session.PlaybackStateCompat
import androidx.core.app.NotificationCompat
import androidx.core.app.NotificationManagerCompat
import androidx.media3.common.AudioAttributes
import androidx.media3.common.C
import androidx.media3.common.MediaItem
import androidx.media3.common.MediaMetadata
import androidx.media3.common.Player
import androidx.media3.exoplayer.ExoPlayer
import androidx.media3.session.MediaSession
import androidx.media3.session.MediaSessionService
import kotlinx.coroutines.*

/**
 * 媒体播放前台服务 — Android 16 核心适配
 * 解决锁屏断播、后台被杀、蓝牙线控失效
 */
class PlaybackService : MediaSessionService() {

    private var player: ExoPlayer? = null
    private var mediaSession: MediaSession? = null
    private val serviceScope = CoroutineScope(Dispatchers.Main + SupervisorJob())

    // 播放状态回调给 WebView
    var onPlaybackEvent: ((String, String?) -> Unit)? = null

    companion object {
        const val CHANNEL_ID = "musicflow_playback"
        const val CHANNEL_NAME = "音乐播放"
        const val NOTIFICATION_ID = 1001

        var instance: PlaybackService? = null
            private set

        // JS 统一调用入口
        fun play(url: String?, title: String?, artist: String?) {
            instance?.playMedia(url, title, artist)
        }

        fun pause() = instance?.player?.pause()
        fun resume() = instance?.player?.play()
        fun seekTo(ms: Long) = instance?.player?.seekTo(ms)
        fun getCurrentPosition(): Long = instance?.player?.currentPosition ?: 0
        fun getDuration(): Long = instance?.player?.duration ?: 0
        fun isPlaying(): Boolean = instance?.player?.isPlaying ?: false
    }

    override fun onCreate() {
        super.onCreate()
        instance = this

        createNotificationChannel()
        createPlayer()
        createMediaSession()
    }

    // ==================== ExoPlayer ====================

    private fun createPlayer() {
        player = ExoPlayer.Builder(this)
            .setAudioAttributes(
                AudioAttributes.Builder()
                    .setContentType(C.AUDIO_CONTENT_TYPE_MUSIC)
                    .setUsage(C.USAGE_MEDIA)
                    .build(),
                true
            )
            .setHandleAudioBecomingNoisy(true)
            .build()
            .apply {
                addListener(object : Player.Listener {
                    override fun onPlaybackStateChanged(state: Int) {
                        updateMediaSessionState()
                        when (state) {
                            Player.STATE_READY -> {
                                startForegroundNotification()
                                onPlaybackEvent?.invoke("ready", "${duration}ms")
                            }
                            Player.STATE_ENDED -> {
                                stopForeground(STOP_FOREGROUND_REMOVE)
                                onPlaybackEvent?.invoke("ended", null)
                            }
                            Player.STATE_BUFFERING -> {
                                onPlaybackEvent?.invoke("buffering", null)
                            }
                            Player.STATE_IDLE -> {
                                onPlaybackEvent?.invoke("idle", null)
                            }
                        }
                    }

                    override fun onIsPlayingChanged(isPlaying: Boolean) {
                        updateMediaSessionState()
                        onPlaybackEvent?.invoke(
                            if (isPlaying) "playing" else "paused",
                            "${currentPosition}"
                        )
                    }
                })
            }
    }

    // ==================== MediaSession ====================

    private fun createMediaSession() {
        val p = player ?: return

        mediaSession = MediaSession.Builder(this, p)
            .setCallback(object : MediaSession.Callback {
                override fun onPlay() = p.play()
                override fun onPause() = p.pause()
                override fun onSeekTo(positionMs: Long) = p.seekTo(positionMs)
                override fun onSkipToNext() = onPlaybackEvent?.invoke("skipNext", null)
                override fun onSkipToPrevious() = onPlaybackEvent?.invoke("skipPrev", null)
                override fun onStop() {
                    p.stop()
                    stopSelf()
                }
            })
            .build()

        setMediaNotificationProvider { _, _ -> buildNotification() }
    }

    override fun onGetSession(controllerInfo: MediaSession.ControllerInfo): MediaSession? {
        return mediaSession
    }

    private fun updateMediaSessionState() {
        val p = player ?: return
        mediaSession?.let { session ->
            session.player.let { sp ->
                // Sync metadata
                if (p.mediaItemCount > 0) {
                    val meta = p.mediaMetadata
                    sp.setMediaMetadata(sp.mediaMetadata.buildUpon()
                        .setTitle(meta.title)
                        .setArtist(meta.artist)
                        .build())
                }
            }
        }
    }

    // ==================== 播放控制 ====================

    fun playMedia(url: String?, title: String?, artist: String?) {
        if (url.isNullOrEmpty()) return

        val p = player ?: return
        val meta = MediaMetadata.Builder()
            .setTitle(title ?: "未知歌曲")
            .setArtist(artist ?: "未知艺术家")
            .build()

        val mediaItem = MediaItem.Builder()
            .setUri(url)
            .setMediaMetadata(meta)
            .build()

        p.setMediaItem(mediaItem)
        p.prepare()
        p.play()

        // 更新 MediaSession 元数据
        mediaSession?.let { session ->
            session.player.setMediaMetadata(meta)
            session.player.playWhenReady = true
        }
    }

    // ==================== 前台通知 ====================

    private fun createNotificationChannel() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val channel = android.app.NotificationChannel(
                CHANNEL_ID,
                CHANNEL_NAME,
                android.app.NotificationManager.IMPORTANCE_HIGH
            ).apply {
                lockscreenVisibility = android.app.Notification.VISIBILITY_PUBLIC
                setShowBadge(false)
            }
            val nm = getSystemService(android.app.NotificationManager::class.java)
            nm.createNotificationChannel(channel)
        }
    }

    private fun buildNotification(): androidx.media3.session.MediaNotification {
        val p = player ?: return androidx.media3.session.MediaNotification(
            NOTIFICATION_ID, android.app.Notification()
        )

        val intent = Intent(this, MainActivity::class.java).apply {
            flags = Intent.FLAG_ACTIVITY_SINGLE_TOP
        }
        val pendingIntent = PendingIntent.getActivity(
            this, 0, intent,
            PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE
        )

        val title = p.mediaMetadata.title?.toString() ?: "MusicFlow"
        val artist = p.mediaMetadata.artist?.toString() ?: ""

        val notification = NotificationCompat.Builder(this, CHANNEL_ID)
            .setContentTitle(title)
            .setContentText(artist)
            .setSmallIcon(android.R.drawable.ic_media_play)
            .setContentIntent(pendingIntent)
            .setOngoing(true)
            .setPriority(NotificationCompat.PRIORITY_HIGH)
            .setStyle(
                androidx.media.app.NotificationCompat.MediaStyle()
                    .setMediaSession(
                        (mediaSession?.sessionCompat as? MediaSessionCompat)?.sessionToken
                    )
                    .setShowActionsInCompactView(0, 1, 2)
            )
            .addAction(
                android.R.drawable.ic_media_previous, "上一首",
                null // handled by MediaSession
            )
            .addAction(
                if (p.isPlaying) android.R.drawable.ic_media_pause
                else android.R.drawable.ic_media_play,
                if (p.isPlaying) "暂停" else "播放",
                null
            )
            .addAction(
                android.R.drawable.ic_media_next, "下一首",
                null
            )
            .build()

        return androidx.media3.session.MediaNotification(NOTIFICATION_ID, notification)
    }

    private fun startForegroundNotification() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            startForeground(
                NOTIFICATION_ID,
                buildNotification().notification,
                ServiceInfo.FOREGROUND_SERVICE_TYPE_MEDIA_PLAYBACK
            )
        } else {
            startForeground(NOTIFICATION_ID, buildNotification().notification)
        }
    }

    // ==================== 生命周期 ====================

    override fun onTaskRemoved(rootIntent: Intent?) {
        // 保留后台播放，不销毁服务
        if (player?.isPlaying == true) {
            // 继续播放
        } else {
            stopSelf()
        }
    }

    override fun onDestroy() {
        serviceScope.cancel()
        instance = null
        player?.release()
        player = null
        mediaSession?.release()
        mediaSession = null
        super.onDestroy()
    }
}